sys/riscv/pmap.c

```diff
@@ -1,168 +1,708 @@
+    WITH_MTX_LOCK (&pmap->mtx) {
+      pt_entry_t *ptep = pmap_lookup_pte(pmap, va);
+      assert(ptep);
+      pt_entry_t pte = *ptep;
+      pte |= set;
+      pte &= ~clr;
+      pmap_write_pte(pmap, ptep, pte, va);
+    }
+  }
 }
 
 bool pmap_is_modified(vm_page_t *pg) {
-  panic("Not implemented!");
+  return pg->flags & PG_MODIFIED;
 }
 
 bool pmap_is_referenced(vm_page_t *pg) {
-  panic("Not implemented!");
+  return pg->flags & PG_REFERENCED;
+}
+
+void pmap_set_modified(vm_page_t *pg) {
+  pg->flags |= PG_MODIFIED;
+  pmap_modify_flags(pg, PTE_D | PTE_W, 0);
 }
 
 void pmap_set_referenced(vm_page_t *pg) {
-  panic("Not implemented!");
+  pg->flags |= PG_REFERENCED;
+  pmap_modify_flags(pg, PTE_A | PTE_V, 0);
 }
 
-void pmap_set_modified(vm_page_t *pg) {
-  panic("Not implemented!");
+bool pmap_clear_modified(vm_page_t *pg) {
+  bool prev = pmap_is_modified(pg);
+  pg->flags &= ~PG_MODIFIED;
+  pmap_modify_flags(pg, 0, PTE_D | PTE_W);
+  return prev;
+}
+
+bool pmap_clear_referenced(vm_page_t *pg) {
+  bool prev = pmap_is_referenced(pg);
+  pg->flags &= ~PG_REFERENCED;
+  pmap_modify_flags(pg, 0, PTE_A | PTE_V);
+  return prev;
 }
 
 int pmap_emulate_bits(pmap_t *pmap, vaddr_t va, vm_prot_t prot) {
-  panic("Not implemented!");
+  paddr_t pa;
+
+  WITH_MTX_LOCK (&pmap->mtx) {
+    if (!pmap_extract_nolock(pmap, va, &pa))
+      return EFAULT;
+
+    pt_entry_t pte = *pmap_lookup_pte(pmap, va);
+
+    if ((prot & VM_PROT_READ) && !(pte & PTE_SW_READ))
+      return EACCES;
+
+    if ((prot & VM_PROT_WRITE) && !(pte & PTE_SW_WRITE))
+      return EACCES;
+
+    if ((prot & VM_PROT_EXEC) && !(pte & PTE_X))
+      return EACCES;
+  }
+
+  vm_page_t *pg = vm_page_find(pa);
+  assert(pg);
+
+  WITH_MTX_LOCK (&pv_list_lock) {
+    /* Kernel non-pageable memory? */
+    if (TAILQ_EMPTY(&pg->pv_list))
+      return EINVAL;
+  }
+
+  pmap_set_referenced(pg);
+  if (prot & VM_PROT_WRITE)
+    pmap_set_modified(pg);
+
+  return 0;
 }
 
-void pmap_activate(pmap_t *pmap) {
-  panic("Not implemented!");
+/*
+ * Physical map management routines.
+ */
+
+static void pmap_setup(pmap_t *pmap) {
+  assert(pmap->pde);
+  pmap->asid = alloc_asid();
+  pmap->satp = SATP_MODE_SV32 | ((paddr_t)pmap->asid << SATP_ASID_S) |
+               (pmap->pde >> PAGE_SHIFT);
+  mtx_init(&pmap->mtx, 0);
+  TAILQ_INIT(&pmap->pte_pages);
+  TAILQ_INIT(&pmap->pv_list);
 }
 
-pmap_t *pmap_lookup(vaddr_t va) {
-  panic("Not implemented!");
+void init_pmap(void) {
+  kernel_pmap.pde = kernel_pde;
+  pmap_setup(&kernel_pmap);
 }
 
-pmap_t *pmap_kernel(void) {
-  panic("Not implemented!");
+pmap_t *pmap_new(void) {
+  pmap_t *pmap = pool_alloc(P_PMAP, M_ZERO);
+  vm_page_t *pg = pmap_pagealloc();
+  pmap->pde = pg->paddr;
+  pmap_setup(pmap);
+
+  /* Install kernel pagetables. */
+  const size_t off = PAGESIZE / 2;
+  WITH_MTX_LOCK (&kernel_pmap.mtx) {
+    memcpy((void *)phys_to_dmap(pmap->pde) + off,
+           (void *)phys_to_dmap(kernel_pde) + off, PAGESIZE / 2);
+  }
```

Comment:
Here is some interesting assumption about how page tables are working on risc-v. Please elaborate :)